### PR TITLE
K8SPS-28: Allow custom config for MySQL

### PR DESCRIPTION
**Pull request by egegunes**
We support three different ways to customize MySQL configuration:

1. `spec.mysql.configuration` field in `cr.yaml`. In this case, the operator will
   create the ConfigMap.
2. Creating a ConfigMap named `<cluster-name>-mysql` that contains the key `my.cnf`
3. Creating a Secret named `<cluster-name>-mysql` that contains the key `my.cnf`

We project ConfigMap and Secret in the same volume. In the entrypoint we
concat these files into MySQL config file (`/etc/my.cnf.d/node.cnf`). By
projecting these two resources, users will be able to split their
customizations. Sensitive values can live in the Secrets while the rest
live in the ConfigMap.

To automatically restart the MySQL pods on configuration change, we
calculate the md5 hash from the value of the `spec.mysql.configuration`
and store it in an annotation in the MySQL StatefulSet's PodSpec.
**This only applies to configuration in cr.yaml. Users need to restart
the MySQL pods if they create/update the configmap themselves.**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/c07505e31a7ab0ebba2ef87ea7e11ad7b1d52029
image: `perconalab/percona-server-mysql-operator:PR-24-c07505e`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/c07505e31a7ab0ebba2ef87ea7e11ad7b1d52029
image: `perconalab/percona-server-mysql-operator:PR-24-c07505e`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/8e06d75d99d9f128d4588d18e8b6fd71d9c0f31a
image: `perconalab/percona-server-mysql-operator:PR-24-8e06d75`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/8e06d75d99d9f128d4588d18e8b6fd71d9c0f31a
image: `perconalab/percona-server-mysql-operator:PR-24-8e06d75`

**Review by delgod — DISMISSED** [review]

**Review by defbin — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/ae14baed47ea9612f763eeaa9bbd8f1c26313e0c
image: `perconalab/percona-server-mysql-operator:PR-24-ae14bae`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | failed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/37730a5cd77783f484e86bfcb3f345b2f4729c05
image: `perconalab/percona-server-mysql-operator:PR-24-37730a5`

File: e2e-tests/tests/config/04-remove-config.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  secretsName: test-secrets
+  sslSecretName: test-ssl
+
+  mysql:
+    image: percona/percona-server:8.0
+    imagePullPolicy: IfNotPresent
+
+    size: 3
+
+    configuration: ""
```
Comment (delgod) [inline e2e-tests/tests/config/04-remove-config.yaml:15]: but how we know that configmap was deleted or cleaned up

File: e2e-tests/tests/init-deploy/01-create-cluster.yaml [hunk 2]
```diff
@@ -1,11 +1,31 @@
...
+  sslSecretName: test-ssl
+  mysql:
+    image: percona/percona-server:8.0
+    imagePullPolicy: Always
 
-    source ../../functions
+    size: 3
+    sizeSemiSync: 0
 
-    get_cr | kubectl apply -n "${NAMESPACE}" -f -
```
Comment (delgod) [inline e2e-tests/tests/init-deploy/01-create-cluster.yaml:11]: idea behind `get_cr` function - have control which MySQL, Orchestrator image is run during the release testing

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/mysql/mysql.go [hunk 3]
```diff
@@ -125,6 +135,43 @@ func StatefulSet(cr *apiv2.PerconaServerForMySQL, initImage string) *appsv1.Stat
...
+							{
+								Name: configVolumeName,
+								VolumeSource: corev1.VolumeSource{
+									Projected: &corev1.ProjectedVolumeSource{
+										Sources: []corev1.VolumeProjection{
+											{
+												ConfigMap: &corev1.ConfigMapProjection{
+													LocalObjectReference: corev1.LocalObjectReference{
+														Name: ConfigMapName(cr),
+													},
```
Comment (defbin) [inline pkg/mysql/mysql.go:147]: indention

**Review by defbin — APPROVED** [review]
lgtm

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | passed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/f6c58f0b9df81cc424435981d81459bdc76e3293
image: `perconalab/percona-server-mysql-operator:PR-24-f6c58f0`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| semi-sync | passed |
| monitoring | passed |
| sidecars | passed |

commit: https://github.com/percona/percona-server-mysql-operator/pull/24/commits/1df077d95cd3c0aa84a5e161d7a4fc3e1a264122
image: `perconalab/percona-server-mysql-operator:PR-24-1df077d`

File: e2e-tests/tests/init-deploy/01-create-cluster.yaml (hunk 2 above)
Comment (egegunes) [reply e2e-tests/tests/init-deploy/01-create-cluster.yaml]: I see your point. I separated the test commit from this PR. Let's merge this and I'll revert `get_cr`

**Review by delgod — DISMISSED** [review]

**Review by delgod — APPROVED** [review]

**Review by defbin — APPROVED** [review]

**Pull request merged**